File: Yixuan/17_01_17/cloest_number_in_sorted_array.py
class Solution:
    # @param {int[]} A an integer array sorted in ascending order
    # @param {int} target an integer
    # @return {int} an integer
    # @author: Egbert Li
    # @link: http://www.lintcode.com/en/problem/closest-number-in-sorted-array/
    def closestNumber(self, A, target):
        if len(A) == 0:
            return 0
        start = 0
        end = len(A) - 1
        while start + 1 < end:
            mid = int(start + (end - start) / 2)
            if target == A[mid]:
                return mid
            elif target < A[mid]:
                end = mid
            else:
                start = mid
        if target - A[start] <= A[end] - target:
            return start
        else:
            return end

File: Yixuan/17_01_17/cloest_number_in_sorted_array.cpp
class Solution {
public:
    /**
     * @param A an integer array sorted in ascending order
     * @param target an integer
     * @return an integer
     * @link: http://www.lintcode.com/en/problem/closest-number-in-sorted-array/
     * @author : Egbert Li
     */
    int closestNumber(vector<int>& A, int target) {
        int length = A.size();
        if (length == 0) {
            return 0;
        }
        int start = 0;
        int end = length - 1;

        // get the closest number, aka the minimum number compare to target
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (target == A[mid]) {
                return mid;
            } else if (target < A[mid]) {
                end = mid;
            } else {
                start = mid;
            }
        }
        if (target - A[start] <= A[end] - target) {
            return start;
        } else {
            return end;
        }
    }
};
